[Merged by Bors] - feat: tactic `gcongr` for "relational congruence"

hrmacbeth opened the pull request:
This PR implements a new tactic, `gcongr`, which applies "relational congruence" rules, reducing a relational goal between a LHS and RHS matching the same pattern to relational subgoals between the differing inputs to the pattern.  For example,
```lean
example {a b x c d : ℝ} (h1 : a + 1 ≤ b + 1) (h2 : c + 2 ≤ d + 2) :
    x ^ 4 * a + c ≤ x ^ 4 * b + d := by
  gcongr
  · linarith
  · linarith
```
This example has the goal of proving the relation `≤` between a LHS and RHS both of the pattern
```
x ^ 4 * ?_ + ?_
```
(with inputs `a`, `c` on the left and `b`, `d` on the right); after the use of `gcongr`, we have the simpler goals `a ≤ b` and `c ≤ d`.

For a sense of the style of argument facilitated by the tactic, [this commit](https://github.com/leanprover-community/mathlib4/commit/9a5e0ce9f9b714fd5383610a12746cded21f030f) (which will be PR'd separately) gives >100 examples of use cases in the existing library.

The tactic's syntax allows for a pattern to be provided explicitly; this is useful if a non-maximal match is desired:
```lean
example {a b c d x : ℝ} (h : a + c + 1 ≤ b + d + 1) :
    x ^ 4 * (a + c) + 5 ≤ x ^ 4 * (b + d) + 5 := by
  gcongr x ^ 4 * ?_ + 5
  linarith
```
This feature is the analogue for general relations of the mathlib3 [`congrm` tactic](https://leanprover-community.github.io/mathlib_docs/tactics.html#congrm).

The "relational congruence" rules used are the library lemmas which have been tagged with the attribute `@[gcongr]`.  For example, the first example constructs the proof term
```lean
add_le_add (mul_le_mul_of_nonneg_left _ (pow_bit0_nonneg x 2)) _
```
using the relational congruence lemmas `add_le_add` and `mul_le_mul_of_nonneg_left`.  In this initial implementation, the `@[gcongr]` tagging has been set up for arithmetic head functions (`+`, `*` etc) and the relations `≤`, `<` and congruence-mod-`n`.

The tactic attempts to discharge side goals to these "relational congruence" lemmas (such as the side goal `0 ≤ x ^ 4` in the above application of `mul_le_mul_of_nonneg_left`) using the tactic `gcongr_discharger`, which wraps `positivity` but can also be extended. Side goals not discharged in this way are left for the user.

[Zulip discussion](https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/New.20tactic.20for.20.22relational.20congruence.22)

Co-authored-by: Mario Carneiro <di.gama@gmail.com>

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

fpvandoorn reviewed (commented):
Looks good, and it seems like a nice tactic to have.

fpvandoorn commented on Mathlib/Tactic/RelCongr/Core.lean:
```diff
@@ -0,0 +1,479 @@
+  for lem in (relCongrExt.getState (← getEnv)).findD (relName, lhsHead) #[] do
+    -- Filter further according to whether the boolean-array of varying/nonvarying arguments of such
+    -- a lemma matches `varyingArgs`.
+    if lem.varyingArgs == varyingArgs then
```
It seems that my expectation is true, which is a design decision we have to be careful with (that if we have a `tt ~ tt` lemma, we also need the corresponding `tt ~ ff` and `ff ~ tt` lemmas).

digama0 replied:
The initial plan was to have this automatically select the next best match and close the extra goals by refl, but @hrmacbeth argued that it should be fine since we don't currently have any cases where there are such holes in the coverage, and it is not a huge ask to keep it that way.

hrmacbeth replied:
Yes ... my opinion: it seems like the correct method to find the "next best" lemma would be rather heavyweight to implement (consider all tagged lemmas whose list-of-booleans is greater than or equal to the list-of-booleans represented by the goal, wrt to the partial order on lists of booleans, and select a lemma which is a minimal element of this set), so a lot of work, considering the current setting when there are only about 20 lemmas tagged anyway.

If the number of tagged lemmas starts to explode, it would certainly make sense to revisit this.

digama0 replied:
As long as we are using equality though for the list-of-bools, we could use it as an additional hash key instead of looping over all of them.

fpvandoorn replied:
I think this is reasonable. It should be clearly documented though.

digama0 commented on Mathlib/Tactic/RelCongr/Core.lean:
```diff
@@ -0,0 +1,479 @@
+    -- and also build a array of booleans according to which arguments `_ ... _` to the head
+    -- function differ between the LHS and RHS
+    (lhsArgs.zip rhsArgs).mapM fun (lhsArg, rhsArg) =>
+      return (none, !(← isDefEq lhsArg rhsArg))
```
We weren't sure about this one. Since this is a congr-like tactic I think we want to be able to prove subgoals where the two sides are semireducible defeq, but maybe it's too early to use it here and we should try this only on unsolved main goals.

hrmacbeth replied:
Should it at least be `withReducibleAndInstances`?  I'm not sure about that vs `isDefEq`, but I assume you want to identify defeq instance arguments?

fpvandoorn replied:
Oh, `withReducibleAndInstances` is better indeed.
I believe that in Lean 3 checking `isDefEq` on some completely different objects can be very expensive with semireducible transparancy, since it may involve unfolding *many* definitions. 
I think that the (nested) `isDefEq` checks in `relAssumption` should probably also be done with `withReducibleAndInstances`.

hrmacbeth replied:
Done!

hrmacbeth commented on Mathlib/Tactic/RelCongr/Core.lean:
```diff
@@ -0,0 +1,479 @@
+    let assum g := g.relAssumption hyps
+    -- Time to actually run the core tactic `Lean.MVarId.relCongr`!
+    let (_, #[]) ← g.relCongr none [] disch assum
+      | throwError "rel failed, cannot prove goal by 'substituting' the listed relationships"
```
This one (`rel`) is strictly weaker than `rel_congr` -- it is actually the version I want for teaching, and in such a context I prefer finishing-only tactics.  I think it's reasonable to discuss whether teaching tactics belong in mathlib (I included it because I thought it might be not just me who wanted to use it for teaching), but if we decide to let it stay, can we preserve my preference on this point?

fpvandoorn replied:
I'm fine with the tactic failing, but I think the error message should show the unsolved goals. I think that is useful even in a teaching tactic, no?

hrmacbeth replied:
Working on it, see question at
https://leanprover.zulipchat.com/#narrow/stream/239415-metaprogramming-.2F-tactics/topic/Alternative.20.22toMessageData.22.20for.20a.20list

hrmacbeth replied:
Done!

hrmacbeth commented on Mathlib/Tactic/RelCongr/Core.lean:
```diff
@@ -0,0 +1,479 @@
+
+def _root_.Lean.MVarId.exact (g : MVarId) (e : Expr) : MetaM Unit := do
+  let .true ← isDefEq (← g.getType) (← inferType e) | failure
+  g.checkNotAssigned `myExact
```
I actually opened a Std PR for this (https://github.com/leanprover/std4/pull/137) but it is blocked on Mario figuring out what to name it :)

hrmacbeth replied:
I added a TODO referencing that PR.

hrmacbeth commented:
> One additional thing I noticed: if there is no matching `rel_congr` lemma, the tactic silently does nothing, but it should fail.
> 
> For example, since there is no `tt ^ tt` lemma for `≤`, it means that the following does nothing (at least at the bottom of `Mathlib.Tactic.RelCongr`)
> 
> ```lean
> example (n k : ℕ) : n ^ k ≤ k ^ n := by 
>   rel_congr
> ```

I think this is the right behaviour, right?  Because in the case of 
```lean
example (n k : ℕ) : 2 * (n ^ k) ≤ 2 * (k ^ n) := by 
  rel_congr
```
you want it to succeed and give the goal `n ^ k ≤ k ^ n`.

The template version does complain if there's a lemma missing, see the tests:
```lean
example {x y : ℕ} (h : f x ≤ f y) : f x ^ 2 ≤ f y ^ 2 := by
  success_if_fail_with_msg
    "rel_congr failed, no @[rel_congr] lemma applies for the template portion f ?a and the relation LE.le"
    (rel_congr (f ?a) ^ 2)
  rel_congr
```

fpvandoorn commented:
I completely agree that the tactic should succeed on `2 * (n ^ k) ≤ 2 * (k ^ n)`. There should just be a bit of logic "if I didn't apply any lemma, fail with a no progress error message".

fpvandoorn commented:
Though this issue might be automatically solved if we follow @digama0's suggestion to key on the list of booleans, since then we will get the error message that no relation has been found (the error message should then explain that this might be because you might need to specialize lemmas to work for when arguments are the same).

fpvandoorn commented on Mathlib/Tactic/RelCongr/Core.lean:
```diff
@@ -265,61 +265,59 @@ partial def _root_.Lean.MVarId.relCongr
   let varyingArgs := tplArgs.map (·.2)
   if varyingArgs.all not then
-    throwError "try refl"
+    throwError "try rfl"
```
even better: you could generate a clickable suggestion with `Std.Tactic.TryThis.addSuggestion`

hrmacbeth replied:
I would prefer not to since there is no guarantee that `rfl` will actually work here -- the suggestion is made whether or not the relation is actually reflexive.  (Actually, I preferred not to have a special case here at all, but Mario thinks it might be helpful.)

mcdoll commented on Mathlib/Tactic/RelCongr/Core.lean:
```diff
@@ -0,0 +1,479 @@
+  -- The core tactic `Lean.MVarId.relCongr` will be run with "assumption tactic"
+  -- consisting of running `Lean.MVarId.relAssumption` (trying a term together with limited
+  -- forward-reasoning on that term) on each nontrivial hypothesis.
+  let assum g := do
```
I have started to factor out the used lemmas, so that one has just a list of maps `Expr -> MVarId -> MetaM Unit` and this can be passed into the `relAssumption` function. Feel free to revert if there is a smarter solution.

hrmacbeth replied:
This is now done, thanks to Mario.

hrmacbeth commented:
> One additional thing I noticed: if there is no matching `rel_congr` lemma, the tactic silently does nothing, but it should fail.

This is now fixed, and I added the test
```lean
example (n k : ℕ) (H : n ^ k + 1 ≤ k ^ n + 1) : n ^ k ≤ k ^ n := by
  success_if_fail_with_msg
    "rel_congr did not make progress"
    (rel_congr)
  linarith
```

hrmacbeth commented:
OK, I think all @fpvandoorn's review comments have been addressed, and it builds and lints.  I'm going to do a global search and replace to rename to `gcongr` (see [naming discussion](https://leanprover.zulipchat.com/#narrow/stream/116395-maths/topic/New.20tactic.20for.20.22relational.20congruence.22)) and then I am satisfied with it!

kmill commented:
Looks good, let's unleash this upon mathlib (and the summer school)!

bors r+
